### ISSUE TITLE
xref-find-references and LSP

### ISSUE THREAD

**Issue by joaotavora** [joaotavora/eglot#147]
Forked from #146. Paging @dgutov and @MaskRay 

I've had a look at https://debbugs.gnu.org/cgi/bugreport.cgi?bug=29619, and I disagree that `xref-find-references` is misdesigned.  I agree with @dgutov there.  So , I think we should lobby for the specification to allow finding references to an arbitrary identifier, without knowing the actual position of that identifier. However if the problem is the useless prompt, we could possibly hack something into its function that is only active in `eglot--managed-mode`.

**Comment by dgutov** [comment]
> So , I think we should lobby for the specification to allow finding references to an arbitrary identifier, without knowing the actual position of that identifier.

I think the specification already allows this. Maybe the documentation is not too great, but you're welcome to return fully qualified identifier names (or whatever other meaningful strings you want) in the identifier completion table, and let the user pick among those.

> However if the problem is the useless prompt

Why do you think it's useless? It's a user preference: if they want to see which identifier is going to be searched for, why not let them? And how is "searching for an arbitrary identifier" going to work without a prompt?

Anyway, there's a variable (mentioned in the debbugs' thread comments) that allows to turn it off.

**Comment by joaotavora** [comment]
> I think the specification already allows this.

Really, where? I meant the LSP specification [here](https://microsoft.github.io/language-server-protocol/specification)

> Why do you think it's useless

Again, it's useless in the face of an LSP specification that doesn't have any mechanism to make use of it.

**Comment by joaotavora** [comment]
@dgutov re-reading your reply, I do believe there was a misunderstanding here.  You probably thought I was talking about xref.el and how it could be improved.  I was actually talking about how LSP should take a hint or two from xref.el.

**Comment by dgutov** [comment]
You are right, sorry.

Not sure how I could help in this discussion, then, but I will follow it with interest.

**Comment by joaotavora** [comment]
Yes, but that doesn't work for references, only definitions, and the query
parameter can't be empty to list every known symbol.

On Nov 6, 2018 20:24, "mkcms" <notifications@github.com> wrote:

I think we should lobby for the specification to allow finding references
to an arbitrary identifier,

Well, there already is workspace/symbol

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/joaotavora/eglot/issues/147#issuecomment-436395928>,
or mute
the thread

**Comment by matklad** [comment]
> I think we should lobby for the specification to allow finding references to an arbitrary identifier, without knowing the actual position of that identifier.

"the spec" here is LSP, right? If it is, I don't think such feature should be included into LSP: known position is crucial to return right results. Consider for example, this:

```
fn foo() {
     let x = 92;
     x;
}

fn bar() {
     let x = 62;
     x + x;
}
```

Although the `x`es in different functions have the same name, they are different things, and have different sets of references. 

"find references by identifier" is better cast as a two phase process: first, you use identifier to find all symbols (definitions). Then, when you find the symbol you actually interested in, you invoke "find references" on it. This can be merged into a single function on the client side, if needed.

**Comment by joaotavora** [comment]
> "find references by identifier" is better cast as a two phase process: first, you use identifier to find all symbols (definitions). Then, when you find the symbol you actually interested in, you invoke "find references" on it. This can be merged into a single function on the client side, if needed.

You make good points.

But in the SLIME Common-Lisp IDE, were xref.el was inspired, the problem is solved differently.  I ask for all the references for "x", get all four in your example, but they are categorized.  And some servers can probably allow you to use "foo::x" to get to reference's of foo's x.  

And you _could_ use the location as just another one of those univocal qualifiers, no problem.  But I think you shouldn't _exclusively_ allow that.

Your two-phase process a good idea though if necessarily slower, I will investigate if it can be done easily with xref.el

**Comment by joaotavora** [comment]
> Your two-phase process a good idea though if necessarily slower, I will investigate if it can be done easily with xref.el

Unfortunately, I just checked and some servers, notably pyls doesn't support `workspace/symbol` (perhaps I'm just using an old version), which is necessary for this strategy to work.

**Comment by MaskRay** [comment]
> allow finding references to an arbitrary identifier

I want to argue that `textDocument/references` at point is a very common operation. Whatever the interface evolves to, that operation should be well supported.

With `workspace/symbol` or `textDocument/documentSymbol`, we can obtain symbols whose names satisfy some condition, and follow that with `textDocument/references`. However, `workspace/symbol` `textDocument/documentSymbol` does not necessary return all symbols and for better user experience it can be the case that local variables are excluded.

For `textDocument/documentSymbol`, the parameter `DocumentSymbolParams` only specifies the document identifier (filename). If you make it return all local variables, it may render badly in imenu and it is very unlikely what a common user expects. But if you exclude local variables from it, the "finding references to an arbitrary identifier" use case cannot work.

**Comment by casouri** [comment]
> But in the SLIME Common-Lisp IDE, were xref.el was inspired, the problem is solved differently. I ask for all the references for "x", get all four in your example, but they are categorized. And some servers can probably allow you to use "foo::x" to get to reference's of foo's x.

IIUC, what you mean is get all possible symbols (instead of ask the user to chose one) and display the references of each unique symbols (with distinction among them), right?

Although for me I would like to chose the symbol that I want to find reference for first (@matklad 's route). But why not just have both? Both approach could be useful in different use cases.

> So , I think we should lobby for the specification to allow finding references to an arbitrary identifier, without knowing the actual position of that identifier.

As discussed above, this feature isn't hard to implement in the client. Now the question is just "should it be in client (xref) or server (lsp server)". I have no idea on this matter, though.

**Comment by joaotavora** [comment]
@monnier and @dgutov, a little help here, pretty please:

I'm having some (well, I'm a complete loss) trouble writing an implementation for `xref-backend-identifier-completion-table`.  I want to use (some kind of, perhaps the most basic kind of) substring completion. This is because the completion-fetching primitive in LSP is this one:

```
(lambda (string)
  (mapcar
   (eglot--lambda ((SymbolInformation) name location) name)
   (jsonrpc-request server :workspace/symbol `(:query ,string))))
```

First of all, this returns no completions for the empty list.  Then if the string is something like `a`, this returns a list of strings list like (one string per line):

```
int main(int argc, char* argv[])
void baz()
void bar()
```
and adding a `in` to make `ain` would make the primitive return only:

```
int main(int argc, char* argv[])
```

deleting the `ain` and typing `ba` would obviously return the last two  which means I could probably partially complete to `void ba`.

In other words, it looks like a substring completion problem (but it's really up to the server how it interprets `query`).

Now, I've tried a lot of combinations of `completion-table-dynamic`, `complete-with-action`, raw completion table handling action and doing the substring completion myself, they all seemed complicated.  I read `minibuffer.el` and all the helpers I need *seem* to be there already, but I just can't get them to work.  Completion "styles" and `completion-substring-try-completion` look particularly promising, but I just can get them to work.

What would you recommend here? Is there a good way to solve this without a lot of brittle code?

**Comment by dgutov** [comment]
> First of all, this returns no completions for the empty list

You probably mean "for the empty string". And if so, that might be unfortunate if you want to utilize some of the existing more advanced styles (e.g. `partial-completion`).

Aside from that, I'm not seeing the problem in your description. How does the actual behavior in the provided examples differ from what you were expecting?

**Comment by joaotavora** [comment]
> You probably mean "for the empty string".

Yes I did.

> And if so, that might be unfortunate if you want to utilize some of the existing more advanced styles (e.g. partial-completion).

Whatever the limitation, I can only work around them because of the LSP spec.

> Aside from that, I'm not seeing the problem in your description. How does the actual behavior in the provided examples differ from what you were expecting?

Oh you thought I had something working already, did you? :) I don't. Those examples were all like "I wish I could make something as basic as this work".

(In truth I already have an implementation for that xref.el GF in master eglot.el, but it's very poor and I'm trying to replace it)

**Comment by dgutov** [comment]
> I can only work around them because of the LSP spec

That's unfortunate, and probably means you can't use the non-basic styles properly because they do fuzzy-matching in Lisp.

> Oh you thought I had something working already, did you? :)

It's how it read. :) Anyway, seems like you could create a new completion style, register it in `completion-styles-alist` (see the examples there) and have it used via `completion-category-defaults`.

It's a non-trivial task (Stefan might help with better advice here), but the implementation will likely be smaller than the existing ones because matching will be performed in the external process.

**Comment by joaotavora** [comment]
> It's a non-trivial task (Stefan might help with better advice here), 

Yup non-trivial it is, at least for me.  I'll let Stefan chime in here or bother him directly.

**Comment by joaotavora** [comment]
Continuing this in #314, where it seems to have picked up some wind in its sails, finally.
